packages/osd-pm/src/utils/validate_dependencies.ts

```diff
@@ -138,21 +178,166 @@ export async function validateDependencies(osd: OpenSearchDashboards, yarnLock:
+        const { object: deps } = cachedManifests.get(project.path);
+        if (deps?.[`${depName}@${range}`]?.version) {
+          installedVersions.add(deps[`${depName}@${range}`].version);
+          installedDepVersionsCache.set(
+            `${project.name}#${depName}`,
+            deps[`${depName}@${range}`].version
+          );
+        } else {
+          log.warning(`Failed to find the installed version for ${depName}@${range}`);
+          // If we cannot read any one of the installed versions of a depName, there is no point in continuing with it
+          installedVersions.clear();
+          desiredRanges.clear();
+          break depLoop;
+        }
+      }
+
+      desiredRanges.set(range, projects);
+    }
+
+    if (installedVersions.size === 0) {
+      // More than one range is used but couldn't get all the installed versions: call out violation
+      violatingSingleVersionDepRanges.set(depName, ranges);
+    } else if (
+      singleVersionResolution === SingleVersionResolution.LOOSE &&
+      installedVersions.size === 1
+    ) {
+      hasIssues = true;
+
+      /* When validating single-version loosely, ignore multiple ranges when they result in the installation of
+       * a single version.
+       */
+      log.info(
+        `Ignored single version requirement for ${depName} as all installations are using v${
+          installedVersions.values().next().value
+        }.`
+      );
+    } else {
+      const sortedInstalledVersion = Array.from(installedVersions).sort(rcompare);
+      const rangePatterns = Array.from(desiredRanges.keys());
+
+      if (singleVersionResolution === SingleVersionResolution.BRUTE_FORCE) {
+        hasIssues = true;
+
+        const suitableVersion = sortedInstalledVersion[0];
+        const suitableRange = `^${suitableVersion}`;
+
+        // Install the version on all projects that have this dep; use the suitable range.
+        for (const { projects } of ranges) {
+          for (const project of projects) {
+            await project.installDependencyVersion(
+              depName,
+              suitableVersion,
+              depName in project.devDependencies,
+              suitableRange
+            );
+          }
+        }
+
+        const conflictingRanges = ranges
+          .map(({ range, projects }) => `${range} => ${projects.map((p) => p.name).join(', ')}`)
+          .join('\n              ');
+        log.warning(dedent`
+
+            [single_version_dependencies] Multiple version ranges for package "${depName}"
+            were found across different package.json files. A version, v${suitableVersion}, was identified as the most recent
+            already installed replacement. All package.json files have been updated to indicate a dependency on \`${depName}@${suitableRange}\`.
+
+            The conflicting version ranges are:
+              ${conflictingRanges}
+          `);
+      } else if (singleVersionResolution === SingleVersionResolution.LOOSE) {
```

Comment:
Get lost here, can we move line 181-325 to a help function like `getViolatingSingleVersionDepRangesByDepRanges`, in that case we can fail fast inside the function.
```javascript
if (installedVersions.size === 0) {
  xxx
  return violatingSingleVersionDepRanges;
}

switch (singleVersionResolution)
case "x": do something
```
I believe it will be easier to understand the complex logic inside this huge block.

Reply:
> in that case we can fail fast inside the function

We are going out of our way to not fail fast. The goal is to be able to show all of the violating dep-range combos.

I can do a `continue` instead of using `else` but then someone else might find that less readable; readability is highly subjective. 

However, I had a real dream of having a 5th mode which does loose + brute-force, and for that I would need to break the logic apart. I will give it a shot and if it works, I will update the PR.

Reply:
I updated this logic to include what I had dreamed. Hope you find it more readable.

Reply:
Thanks @AMoo-Miki .As for me it looks much easier to under as we compare the Resolution and only the Resolution in the first level. One suggestion is maybe we need to add unit test for all these `if else` branches.

Reply:
I will have to think of a way to test these. They depend on what Yarn does to complete so <fingers crossed>.